src/sync.rs

```diff
@@ -0,0 +1,95 @@
+#[cfg(feature = "parking_lot")]
+mod parking_lot_mutex {
+    pub(crate) use parking_lot::{MutexGuard, WaitTimeoutResult};
+    use std::time::{Duration, Instant};
+
+    pub(crate) struct Mutex<T>(parking_lot::Mutex<T>);
+
+    impl<T> Mutex<T> {
+        #[inline]
+        pub fn new(t: T) -> Self {
+            Self(parking_lot::Mutex::new(t))
+        }
+
+        #[inline]
+        pub fn lock(&self) -> MutexGuard<T> {
+            self.0.lock()
+        }
+    }
+
+    pub(crate) struct Condvar(parking_lot::Condvar);
+
+    impl Condvar {
+        #[inline]
+        pub fn new() -> Self {
+            Self(parking_lot::Condvar::new())
+        }
+
+        #[inline]
+        pub fn notify_one(&self) {
+            self.0.notify_one();
+        }
+
+        #[inline]
+        pub fn wait_until<'a, T>(
+            &self,
+            mut mutex_guard: MutexGuard<'a, T>,
+            timeout: Duration,
+        ) -> (MutexGuard<'a, T>, WaitTimeoutResult) {
+            let end = Instant::now() + timeout;
```

Comment:
This is not correct. The deadline needs to apply outside the looped calls to wait.

Reply:
> This is not correct. The deadline needs to apply outside the looped calls to wait.

Ow, right. Next logic must be ok:
```rust
        pub fn wait_until<'a, T>(
            &self,
            mutex_guard: MutexGuard<'a, T>,
            timeout: Instant,
        ) -> (MutexGuard<'a, T>, WaitTimeoutResult) {
            let timeout = timeout
                .checked_duration_since(Instant::now())
                .unwrap_or_default();

            self.0
                .wait_timeout(mutex_guard, timeout)
                .unwrap_or_else(PoisonError::into_inner)
        }
```